### DISCUSSION TITLE
Global or Default Namespace Statements

### DISCUSSION THREAD

**Discussion by Turpy** [dotnet/csharplang#4908] [Language Ideas]
Similar to the new global using statements, it would be nice if there was an option for a global namespace that could be set for a project so that if a namespace was not declared in a file then it would default to this global namespace. 

If a namespace was provided then it would override the globally set namespace.

Also, it would be nice if there was a directory structure for source files underneath the top level directory for the project then the name of the directory can be appended automatically to the project level global namespace - again if none are provided in that file.

**Comment by CyrusNajmabadi** [comment]
> then the name of the directory can be appended automatically to the project level global namespace - again if none are provided in that file.

Just an FYI, but we are very unlikely to do either of these as we are generally loathe to have the project file have impact on the meaning of code, or how the metadata for symbols is generated.

**Reply by jnm2** [reply]
I just discovered that the SDK will generate default global usings unless you opt out in csproj, and we'll also be able to add our own global usings in csproj (which I like better than starting a new .cs file). Both of these things change the meaning of code.

**Reply by RikkiGibson** [reply]
It seems like the suggestion is to have the global namespace declared in C# source, perhaps in the same place that the global usings are declared.

**Reply by Turpy** [reply]
Yes, this idea is based off of global usings. It seems like the natural progression would be to also allow a global namespace statement as well in the same location.

I need to look more at how the namespace is generated when using top-level programming in the main file, but I would like this global namespace to be used for it too.

**Reply by RikkiGibson** [reply]
I think top-level statements aren't allowed to be contained in a namespace, and any members synthesized for the top-level statements are contained in the global namespace.

**Reply by HakamFostok** [reply]
@CyrusNajmabadi 
>  but we are very unlikely to do either of these as we are generally loathe to have the project file have impact on the meaning of code

"Implicit usings" did that exactly, and I think Implicit usings was one of the greatest things that happened to C#, IMO.<br/>
IMO, this idea is as good as an "implicit usings", and it could be natural progression for the "Implicit usings" <br/>
And it will align well with the theme that C# took to make it easier for the newcomers. <br/>
We do not have to teach them about the namespace at the beginning, this could be taught later.

**Comment by jnm2** [comment]
~~The first ask is the same as https://github.com/dotnet/csharplang/discussions/4902, I think.~~ The second ask is almost the same as https://github.com/dotnet/csharplang/discussions/2908.

**Reply by Turpy** [reply]
I'm not asking for an option to select a namespace generation algorithm. I would like this to be done the same as global using statements and have the namespaces be declared explicitly.

I'm less interested in the second part, but since there are code suggestions to make namespaces match directory structure in many projects, depending on editorconfig, IDE settings, and analyzers, it may be nice to have an option to have the global namespaces appended with directory. However I know that would be controversial since I work with projects that are strongly opinionated on both having a single namespace per project or having multiple they reflect the directory structure.

I'm not suggesting that any of these be enabled by default. The code in the file should also always have precedence.

**Reply by jnm2** [reply]
You're right, I misunderstood #4902.

**Comment by jnm2** [comment]
With your suggestion, if I had a default namespace specified in csproj called `NotInventedHere`, then my class that are directly in the namespace `NotInventedHere` would be in files with no `namespace`  keyword, and my classes that are in the `NotInventedHere.Foo` namespace would have to specify `namespace NotInventedHere.Foo`? The difference between these two files seems like it could potentially be a bit confusing, and not all that rare given the prevalence of subnamespaces pushed by the IDE if you use subfolders.

This could also come up in a single file:
```cs
class A { }

namespace NotInventedHere.Foo
{
    class B { }
}
```

What would you say is gained by that default namespace then?

https://github.com/dotnet/csharplang/issues/137 is championed, giving the ability to write `namespace NotInventedHere;` at the top of the file. The gain with this one is reduced indentation, and the gain is achieved even in the (common?) scenario of nested namespaces.

**Reply by Turpy** [reply]
Essentially, I am requesting that if someone writes a global or default namespace statement then all classes that would be in the "global" would instead be in the namespace specified by the user. Everything else would behave the same.

So in your example, class A today would be in the "global" namespace and class B would be in the "NotInventedHere.Foo" namespace. If there is a global namespace declaration (structured like the global usings) of "MyGlobalNamespace" then class in your example would be in it but class B would still be the in the same namespace declared in the code.

**Reply by Turpy** [reply]
But I am also aware of what you are staying. I am just not sure what the best approach would be for sub namespaces. I don't know if would be better for them to be generated somehow by having some additional options or just leave it to where they have to be declared explicitly.

**Comment by HakamFostok** [comment]
That is a great idea, and I hope this will be implemented

**Reply by Joe4evr** [reply]
Congratulations, [it shipped 3 months ago](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-10#global-using-directives).

**Reply by MgSam** [reply]
@Joe4evr No, it didn't.

**Reply by HakamFostok** [reply]
@Joe4evr, it is not, please read the suggestion again
